fix  #1079 - Add support for Vue 2.7

Florent-Bouisset opened the pull request:
## Summary

Add support for vue 2.7

Allow to unit test for 2.7 with vue-demi

fixes #1079


Even if vue 2.7 doesn't need `vue-template-compiler` itself
it's still needed because of `@vue/test-utils`  and need to match the current vue version, so it is why I added this behavior to `swap-vue.js`


https://blog.vuejs.org/posts/vue-2-7-naruto.html
> Upgrade vue to ^2.7.0. You can also remove vue-template-compiler from the dependencies - it is no longer needed in 2.7.
> 
> Note: if you are using @vue/test-utils, you may need to keep it in the dependencies for now, but this requirement will also be lifted in a new release of test utils.

fixes #1079



## Metadata

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [ ] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] I have read the [Contribution Guides](https://github.com/vuelidate/vuelidate/blob/master/.github/CONTRIBUTING.md#pull-request-guidelines)
- [x] It's submitted to the `next` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuelidate/vuelidate/blob/master/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

Commit:
upgrading vue-demi to 0.13.4

Commit:
add test for vue2.7

Commit:
change vue-template-compiler version

Commit:
fix unit test for vue 2.7

Commit:
cleanup

dobromir-hristov commented on package.json:
```diff
@@ -13,7 +13,7 @@
+    "vue-demi": "^0.13.4"
   },
   "optionalDependencies": {
     "@vue/composition-api": "^1.3.1"
```
Should we update the composition api version too?

Florent-Bouisset replied:
I don't think it's necessary ?

dobromir-hristov commented on packages/vuelidate/src/index.js:
```diff
@@ -55,7 +55,7 @@ export function useVuelidate (validations, state, globalConfig = {}) {
     removeValidationResultsFromParent
   } = instance
-    ? nestedValidations({ $scope, instance })
+    ? nestedValidations({ $scope, instance: instance.proxy })
```
You sure this is compatible with 2 2.7 and 3 ? I remember I had issues with the `getCurrentInstance` function a few times

dobromir-hristov replied:
Actually why do we even need to do this? We just attach something to the instance, not sure we need to pass the `ComponentPublicInstance`

Florent-Bouisset replied:
I made test with a real app only for 2.7, I trusted the units test for 2.6 and 3 but I'm going to try it

the function `getCurrentInstance` has different signature with 2.6 / 2.7 / 3

// 2.6  unofficial plugin @vue/composition-api
```js
declare function getCurrentInstance(): ComponentInternalInstance | null;
```

// 2.7
```js
export function getCurrentInstance(): { proxy: Vue } | null 
```

// 3
```js
export declare const getCurrentInstance: () => ComponentInternalInstance | null;
```

According to what has been said in this discussion:
https://github.com/vuejs/vue/issues/12596#issuecomment-1173269807

If I understand it right, we should NOT use the internal Instance but the public instance, which can be accessed by  `getCurrentInstance().proxy`

Maybe I'm wrong let me know what you understood about it

dobromir-hristov replied:
Great find, let's keep it that way then. I mainly was not sure if vue 2 would have proxy, so was worried not to regress.

robertmoura replied:
I'm running Vue 3 and proxy doesn't have a `uid` on it. Means that the validations are registered as `_vuelidate_undefined`. Causing issues with multiple child validations.
```
$registerAs = `_vuelidate_${uid}`
```

Florent-Bouisset replied:
Sorry for that :( 

uid is not on the `ComponentPublicInstance` but on the  `ComponentInternalInstance` on vue3 

we can do as before by accessing the internal instance with `instance.$.uid`
```js
    const uid = instance?.$?.uid || instance._uid
    $registerAs = `_vuelidate_${uid}`
```

but as said before accessing internal instance is not recommended, 
maybe we should create our own uid ? I don't think we need to have the same as the component one.

any thoughts @dobromir-hristov ?

dobromir-hristov commented on scripts/swap-vue.js:
```diff
@@ -2,8 +2,15 @@ const fs = require('fs')
 const path = require('path')
 
 const Vue2 = path.join(__dirname, '../node_modules/vue2')
+// eslint-disable-next-line camelcase
```
you can disable it for the entire file :)

dobromir-hristov reviewed (commented):
This looks great, thank you so much! I saw all the tests have passed in CI, just checking to see if it works in real world too.

I am having some issues with my local, so I cant test it right now.

Commit:
code review

Commit:
refactor: :recycle: use .proxy only once at variable initialization

we need to use .proxy to access to the public properties and not the internal properties
https://github.com/vuejs/vue/issues/12596#issuecomment-1173269807

Commit:
docs: :memo: follow the official recommandation about getCurrentInstance

https://github.com/vuejs/vue/issues/12596#issuecomment-1173269807

Commit:
fix: :bug: fix race condition on rename due to async method

ademyalcin27 commented:
Is there any update here?, We are waiting for this development?

dobromir-hristov commented:
I think we are good. I will test this Saturday and cut a release :) thank you so for the patience

dobromir-hristov reviewed (approved):
amazing work :)
